{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sys import path\n",
    "path.append('../src')\n",
    "from NLEval import graph, valsplit, label, model\n",
    "from sklearn.metrics import roc_auc_score as auroc\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load graph and lablset collection\n",
    "data_path = '../data/' # path to data\n",
    "# load graph and labelset collection\n",
    "g = graph.DenseGraph.DenseGraph.from_edglst(data_path \\\n",
    "    + 'networks/STRING-EXP.edg', weighted=True, directed=False)\n",
    "lsc = label.LabelsetCollection.SplitLSC.from_gmt(data_path + 'labels/KEGGBP.gmt')\n",
    "\n",
    "# initialize models\n",
    "SL_A = model.SupervisedLearning.LogReg(g, penalty='l2', solver='lbfgs')\n",
    "LP_A = model.LabelPropagation.LP(g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                  Name  Size  Eff size\n",
      "0                                      KEGG_CELL_CYCLE   128       118\n",
      "1        KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION    44        40\n",
      "2                      KEGG_TGF_BETA_SIGNALING_PATHWAY    86        79\n",
      "3                           KEGG_WNT_SIGNALING_PATHWAY   151       137\n",
      "4                         KEGG_GLYCEROLIPID_METABOLISM    49        38\n",
      "5                         KEGG_BETA_ALANINE_METABOLISM    22        20\n",
      "6                          KEGG_GLUTATHIONE_METABOLISM    50        42\n",
      "7                              KEGG_TASTE_TRANSDUCTION    52        25\n",
      "8                                 KEGG_DNA_REPLICATION    36        34\n",
      "9                      KEGG_CARDIAC_MUSCLE_CONTRACTION    80        66\n",
      "10                          KEGG_RIBOFLAVIN_METABOLISM    16        12\n",
      "11                       KEGG_ECM_RECEPTOR_INTERACTION    84        75\n",
      "12                KEGG_DRUG_METABOLISM_CYTOCHROME_P450    72        49\n",
      "13      KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION    44        41\n",
      "14              KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY   108        98\n",
      "15                        KEGG_SPHINGOLIPID_METABOLISM    40        30\n",
      "16                      KEGG_PENTOSE_PHOSPHATE_PATHWAY    27        24\n",
      "17        KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS    22        20\n",
      "18                             KEGG_RETINOL_METABOLISM    64        43\n",
      "19                           KEGG_STEROID_BIOSYNTHESIS    17        15\n",
      "20   KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_...    26        13\n",
      "21                 KEGG_GLYCEROPHOSPHOLIPID_METABOLISM    77        59\n",
      "22                          KEGG_PYRIMIDINE_METABOLISM    98        87\n",
      "23                           KEGG_LONG_TERM_DEPRESSION    70        58\n",
      "24                                     KEGG_PROTEASOME    48        39\n",
      "25                KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM    34        30\n",
      "26   KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITI...    22        14\n",
      "27                                   KEGG_GAP_JUNCTION    90        82\n",
      "28                          KEGG_N_GLYCAN_BIOSYNTHESIS    46        36\n",
      "29                                    KEGG_ENDOCYTOSIS   183       155\n",
      "30        KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION    86        80\n",
      "31                           KEGG_HISTIDINE_METABOLISM    29        25\n",
      "32      KEGG_ALDOSTERONE_REGULATED_SODIUM_REABSORPTION    42        40\n",
      "33              KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY    75        68\n",
      "34                   KEGG_STEROID_HORMONE_BIOSYNTHESIS    55        33\n",
      "35     KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS    11         9\n",
      "36                         KEGG_ERBB_SIGNALING_PATHWAY    87        84\n",
      "37         KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM    24        19\n",
      "38                   KEGG_DORSO_VENTRAL_AXIS_FORMATION    25        21\n",
      "39                          KEGG_O_GLYCAN_BIOSYNTHESIS    30        19\n",
      "40   KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SE...    15         8\n",
      "41   KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_B...    25        20\n",
      "42                          KEGG_TRYPTOPHAN_METABOLISM    40        32\n",
      "43                KEGG_LIMONENE_AND_PINENE_DEGRADATION    10         9\n",
      "44              KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS    16        12\n",
      "45                                 KEGG_TIGHT_JUNCTION   134       111\n",
      "46                                 KEGG_OOCYTE_MEIOSIS   114       106\n",
      "47                    KEGG_ARACHIDONIC_ACID_METABOLISM    58        43\n",
      "48                KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY    79        69\n",
      "49                       KEGG_PHENYLALANINE_METABOLISM    18        14\n",
      "50                    KEGG_SELENOAMINO_ACID_METABOLISM    26        21\n",
      "51                        KEGG_NOTCH_SIGNALING_PATHWAY    47        40\n",
      "52                               KEGG_ABC_TRANSPORTERS    44        36\n",
      "53                       KEGG_RENIN_ANGIOTENSIN_SYSTEM    17        16\n",
      "54                              KEGG_PURINE_METABOLISM   159       139\n",
      "55             KEGG_CYSTEINE_AND_METHIONINE_METABOLISM    34        29\n",
      "56            KEGG_COMPLEMENT_AND_COAGULATION_CASCADES    69        61\n",
      "57                      KEGG_ONE_CARBON_POOL_BY_FOLATE    17        14\n",
      "58                        KEGG_CIRCADIAN_RHYTHM_MAMMAL    13        13\n",
      "59                                 KEGG_PROTEIN_EXPORT    24        22\n",
      "60                                       KEGG_LYSOSOME   121       106\n",
      "61                    KEGG_AMINOACYL_TRNA_BIOSYNTHESIS    41        37\n",
      "62                         KEGG_ETHER_LIPID_METABOLISM    33        23\n",
      "63                     KEGG_HEMATOPOIETIC_CELL_LINEAGE    88        78\n",
      "64                      KEGG_OXIDATIVE_PHOSPHORYLATION   135       115\n",
      "65                        KEGG_REGULATION_OF_AUTOPHAGY    35        26\n",
      "66   KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION    48        35\n",
      "67                                 KEGG_RNA_POLYMERASE    29        25\n",
      "68                           KEGG_BASE_EXCISION_REPAIR    35        33\n",
      "69                       KEGG_LINOLEIC_ACID_METABOLISM    29        20\n",
      "70                                KEGG_MISMATCH_REPAIR    23        22\n",
      "71                      KEGG_INSULIN_SIGNALING_PATHWAY   137       130\n",
      "72                    KEGG_CHEMOKINE_SIGNALING_PATHWAY   190       165\n",
      "73                                  KEGG_MELANOGENESIS   102        94\n",
      "74                       KEGG_HOMOLOGOUS_RECOMBINATION    28        26\n",
      "75                          KEGG_P53_SIGNALING_PATHWAY    69        64\n",
      "76                       KEGG_OTHER_GLYCAN_DEGRADATION    16        15\n",
      "77               KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS    97        87\n",
      "78                                  KEGG_AXON_GUIDANCE   129       117\n",
      "79      KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY   137       103\n",
      "80           KEGG_PORPHYRIN_AND_CHLOROPHYLL_METABOLISM    41        30\n",
      "81                KEGG_ARGININE_AND_PROLINE_METABOLISM    54        51\n",
      "82                          KEGG_PROPANOATE_METABOLISM    33        31\n",
      "83                           KEGG_BUTANOATE_METABOLISM    34        28\n",
      "84                           KEGG_GALACTOSE_METABOLISM    26        21\n",
      "85                            KEGG_NITROGEN_METABOLISM    23        19\n",
      "86                KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS    15        15\n",
      "87           KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION   118        96\n",
      "88   KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SU...    26        12\n",
      "89        KEGG_PROXIMAL_TUBULE_BICARBONATE_RECLAMATION    23        22\n",
      "90                         KEGG_VEGF_SIGNALING_PATHWAY    76        66\n",
      "91                                     KEGG_PEROXISOME    78        75\n",
      "92           KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY   102        84\n",
      "93                  KEGG_DRUG_METABOLISM_OTHER_ENZYMES    51        36\n",
      "94                              KEGG_SULFUR_METABOLISM    13        10\n",
      "95                 KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS    16        12\n",
      "96                         KEGG_MTOR_SIGNALING_PATHWAY    52        48\n",
      "97     KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM    32        31\n",
      "98                         KEGG_GNRH_SIGNALING_PATHWAY   101        92\n",
      "99                                      KEGG_APOPTOSIS    88        81\n",
      "100                                   KEGG_SPLICEOSOME   128       114\n",
      "101                             KEGG_ADHERENS_JUNCTION    75        68\n",
      "102                    KEGG_NUCLEOTIDE_EXCISION_REPAIR    44        41\n",
      "103                    KEGG_HEDGEHOG_SIGNALING_PATHWAY    56        52\n",
      "104                     KEGG_CALCIUM_SIGNALING_PATHWAY   178       164\n",
      "105           KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION    89        46\n",
      "106                    KEGG_JAK_STAT_SIGNALING_PATHWAY   155       142\n",
      "107                        KEGG_LONG_TERM_POTENTIATION    70        66\n",
      "108                KEGG_NEUROTROPHIN_SIGNALING_PATHWAY   126       118\n",
      "109            KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM    10         8\n",
      "110  KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SU...    15         6\n",
      "111       KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM    16        16\n",
      "112           KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY    62        56\n",
      "113                 KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY    56        40\n",
      "114                           KEGG_PYRUVATE_METABOLISM    40        37\n",
      "115                    KEGG_GLYCOLYSIS_GLUCONEOGENESIS    62        55\n",
      "116                           KEGG_TYROSINE_METABOLISM    42        35\n",
      "117                    KEGG_NON_HOMOLOGOUS_END_JOINING    14        12\n",
      "118                                      KEGG_RIBOSOME    88        82\n",
      "119   KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM    44        40\n",
      "120                  KEGG_CELL_ADHESION_MOLECULES_CAMS   134        99\n",
      "121                 KEGG_STARCH_AND_SUCROSE_METABOLISM    52        38\n",
      "122         KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM    76        65\n",
      "123                KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS   138       129\n",
      "124                           KEGG_FOLATE_BIOSYNTHESIS    11        11\n",
      "125                 KEGG_GLYCOSAMINOGLYCAN_DEGRADATION    21        10\n",
      "126         KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY    71        57\n",
      "127                               KEGG_RNA_DEGRADATION    59        51\n",
      "128                        KEGG_PPAR_SIGNALING_PATHWAY    69        63\n",
      "129     KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT    38        36\n",
      "130                       KEGG_CITRATE_CYCLE_TCA_CYCLE    32        30\n",
      "131               KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY    67        62\n",
      "132            KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION   115       102\n",
      "133      KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS    28        20\n",
      "134   KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GLOBO_SERIES    14         7\n",
      "135       KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM    31        28\n",
      "136                         KEGG_FATTY_ACID_METABOLISM    42        40\n",
      "137                   KEGG_BASAL_TRANSCRIPTION_FACTORS    36        30\n",
      "138                            KEGG_LYSINE_DEGRADATION    44        40\n"
     ]
    }
   ],
   "source": [
    "# display label info\n",
    "pd.options.display.max_rows = 999\n",
    "df = pd.DataFrame()\n",
    "df['Name'] = lsc.labelIDlst\n",
    "df['Size'] = [len(lsc.getLabelset(i)) for i in lsc.labelIDlst]\n",
    "df['Eff size'] = [len(lsc.getLabelset(i) & set(g.IDmap.lst)) for i in lsc.labelIDlst]\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def example(label_index, numbers_of_top_genes, model):\n",
    "    # get labelset ID\n",
    "    labelID = lsc.labelIDlst[label_index]\n",
    "    print(labelID)\n",
    "    \n",
    "    # get positive and negative samples\n",
    "    positive_set = lsc.getLabelset(labelID)\n",
    "    negative_set = lsc.getNegative(labelID)\n",
    "    \n",
    "    # train and get genome wide prediction scores\n",
    "    score_dict = model.predict(positive_set, negative_set)\n",
    "    \n",
    "    # print top ranked genes and its intersection with known ones\n",
    "    top_list = sorted(score_dict, key=score_dict.get, reverse=True)[:numbers_of_top_genes]\n",
    "    intersection = list(set(top_list) & positive_set)\n",
    "    print(\"Top %d genes: %s\" % (numbers_of_top_genes, repr(top_list)))\n",
    "    print(\"Known genes in top %d: %s\" % (numbers_of_top_genes, repr(intersection)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEGG_ENDOCYTOSIS\n",
      "Top 50 genes: ['7157', '3320', '10273', '2064', '5710', '2534', '8850', '10075', '2547', '24148', '9351', '9040', '6426', '5052', '9092', '6720', '7525', '6389', '6446', '51639', '8766', '1453', '6727', '2308', '5170', '637', '10253', '5327', '9525', '3486', '26088', '330', '10193', '5584', '1643', '3561', '6135', '5717', '2956', '6154', '5528', '8650', '51529', '8867', '4773', '27', '6152', '374291', '10580', '64223']\n",
      "Known genes in top 50: ['3561', '8766', '5584', '9525', '10193']\n"
     ]
    }
   ],
   "source": [
    "example(29, 50, SL_A)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
